llvm/lib/Target/DirectX/DXILShaderFlags.cpp

```diff
@@ -81,16 +101,31 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
   OS << ";\n";
 }
 
-/// Return the shader flags mask of the specified function Func.
-const ComputedShaderFlags &
-ModuleShaderFlags::getFunctionFlags(const Function *Func) const {
+auto ModuleShaderFlags::getFunctionShaderFlagInfo(const Function *Func) const {
   const auto Iter = llvm::lower_bound(
       FunctionFlags, Func,
       [](const std::pair<const Function *, ComputedShaderFlags> FSM,
          const Function *FindFunc) { return (FSM.first < FindFunc); });
   assert((Iter != FunctionFlags.end() && Iter->first == Func) &&
          "No Shader Flags Mask exists for function");
-  return Iter->second;
+  return Iter;
+}
+
+/// Merge mask NewSF to that of Func, if different.
+/// Return true if mask of Func is changed, else false.
+bool ModuleShaderFlags::mergeFunctionShaderFlags(
+    const Function *Func, const ComputedShaderFlags NewSF) {
+  const auto FuncSFInfo = getFunctionShaderFlagInfo(Func);
+  if ((FuncSFInfo->second & NewSF) != NewSF) {
+    const_cast<ComputedShaderFlags &>(FuncSFInfo->second).merge(NewSF);
```

Comment:
This presumes that the merge is a simple bitwise OR operation and you can use this `&` to detect a change.  It also uses `&` operator on the `ComputedShaderFlags` type, which I assume works by triggering an implicit cast to `uint64_t`.  I believe it would be cleaner to allow for a copy of the `ComputedShaderFlags` and to compare before and after merge to detect a change.

As a side note, we probably don't need to have all the shader flags defined as bitfield entries with one-by-one translation from bitfields to uint64_t. I certainly get that bitfields don't have a guaranteed layout in C++, but we could instead use an explicit `uint64_t` field with an enum defining the (DxilModule) bit shift for each flag (plus generated set/get accessors).  If we did that, operations such as this would be more efficient, since it doesn't have to go through multiple translations between individual bits and combined `uint64_t` values each time.  Copying and merging would be made trivial as well.

Reply:
This is very much an abuse of the user-defined conversion operator... maybe we should remove that to prevent future abuses.

It seems like a better approach here since you're looking to compare equality is to just implement `operator==`.

Reply:
Wait... I'm clearly not thinking straight because that last comment doesn't fully make sense. I do think we should probably get rid of the implicit conversion operator. Clearly `operator==` isn't the right approach.

For flag merging, we maybe don't even need to check it conditionally though. This should all be really well optimized straight-line code to merge flags, so the branch is probably unnecessary.

Reply:
Changed `merge()` to operate on `ComputeShaderFlags` type thereby not relying on implicit cast to `uint64_t`.

Updated `mergeFunctionShaderFlags()` with the check for shader flag value change.

However, testing for equality still implicitly casts to `uint64_t`. Getting rid of implicit conversion operator requires changes to other access functions such as `getFeatureFlags()` etc., that use the conversion. I'll plan to delete implicit conversion operator in a follow-on PR. I think this implies that `operator==` would need to be implemented. Created an [issue](https://github.com/llvm/llvm-project/issues/118570) for this.